src/lib/datatypes/sol-memdesc.c

```diff
@@ -0,0 +1,808 @@
+        if (!desc->structure_members) {
+            SOL_WRN("desc=%p is SOL_MEMDESC_TYPE_STRUCTURE but does not provide structure_members", desc);
+            return -EINVAL;
+        }
+
+        return copy_structure(desc, mem, ptr_content);
+    } else if (desc->type == SOL_MEMDESC_TYPE_ARRAY) {
+        if (!desc->array_item) {
+            SOL_WRN("desc=%p is SOL_MEMDESC_TYPE_ARRAY but does not provide array_item", desc);
+            return -EINVAL;
+        }
+
+        return copy_array(desc, mem, ptr_content);
+    }
+
+    memcpy(mem, ptr_content, sol_memdesc_get_size(desc));
+    return 0;
+}
+
+static int
+copy_content(const struct sol_memdesc *desc, const void *src_memory, void *dst_memory)
+{
+    if (desc->ops && desc->ops->copy)
+        return desc->ops->copy(desc, src_memory, dst_memory);
+
+    return set_content(desc, dst_memory, src_memory);
+}
+
+static int
+copy_structure(const struct sol_memdesc *desc, void *memory, const void *ptr_content)
+{
+    const struct sol_memdesc_structure_member *itr;
+    const void *src_mem;
+
+    SOL_MEMDESC_FOREACH_STRUCTURE_MEMBER_MEMORY(desc, itr, ptr_content, src_mem) {
+        void *dst_mem;
+        int r;
+
+        dst_mem = sol_memdesc_get_structure_member_memory(desc, itr, memory);
+        if (!dst_mem)
+            return -errno;
+        r = copy_content(&itr->base, src_mem, dst_mem);
+        if (r < 0)
+            return r;
+    }
+
+    return 0;
+}
+
+static int compare_content(const struct sol_memdesc *desc, const void *a_memory, const void *b_memory);
+
+static int
+compare_structure(const struct sol_memdesc *desc, const void *a_memory, const void *b_memory)
+{
+    const struct sol_memdesc_structure_member *itr;
+    const void *a_mem;
+
+    errno = 0;
+    SOL_MEMDESC_FOREACH_STRUCTURE_MEMBER_MEMORY(desc, itr, a_memory, a_mem) {
+        const void **b_mem;
+        int r;
+
+        b_mem = sol_memdesc_get_structure_member_memory(desc, itr, b_memory);
+        r = compare_content(&itr->base, a_mem, b_mem);
+        if (r != 0 || errno)
+            return r;
+    }
+
+    return 0;
+}
+
+static int
+compare_array(const struct sol_memdesc *desc, const void *a_memory, const void *b_memory)
+{
+    const void *a_item, *b_item;
+    ssize_t idx, len, a_len, b_len;
+
+    a_len = sol_memdesc_get_array_length(desc, a_memory);
+    if (a_len < 0) {
+        errno = -a_len;
+        return 0;
+    }
+    b_len = sol_memdesc_get_array_length(desc, b_memory);
+    if (b_len < 0) {
+        errno = -b_len;
+        return 0;
+    }
+
+    len = sol_util_min(a_len, b_len);
+    SOL_MEMDESC_FOREACH_ARRAY_ELEMENT_IN_RANGE(desc, a_memory, 0, len, idx, a_item) {
+        int r;
+
+        b_item = sol_memdesc_get_array_element(desc, b_memory, idx);
+        if (!b_item)
+            return 0;
+
+        r = compare_content(desc->array_item, a_item, b_item);
+        if (r != 0 || errno)
+            return r;
+    }
+
+    if (idx < len) /* loop failed */
```

Comment:
I guess you could do like ``copy_array()`` does here: ``r = errno ? -errno : -EINVAL;``.